cloudferry/lib/utils/remote.py

```diff
@@ -104,11 +100,92 @@ def tmpdir(self, prefix='cloudferry'):
             if path is not None:
                 self.sudo('rm -rf {path}', path=path)
 
+    @contextlib.contextmanager
+    def open_file(self, filename, mode='r'):
+        transport = self.client.get_transport()
+        sftp_client = paramiko.SFTPClient.from_transport(
+            transport, window_size=WINDOW_SIZE)
+        with sftp_client:
+            yield sftp_client.open(filename, mode, 512 * 1024 * 1024)
+
+    def _connect_through_gateway(self, host, port, gateway):
+        # TODO: connection_attempts
+        if gateway is None:
+            return None
+        client = paramiko.SSHClient()
+        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        client.connect(
+            hostname=gateway.hostname,
+            port=gateway.port,
+            username=gateway.username,
+            password=gateway.password,
+            pkey=self._create_pkey(gateway.private_key),
+            sock=self._connect_through_gateway(
```

Comment:
It done this way in order to connect to controller through several gateways (e.g. first through jump host, then through fuel master).